### PR TITLE
JW-256: KSync implementation for Windows

### PR DESCRIPTION
**Pull request by jablonskim**

### CODE REVIEW

File: src/ksync/ksync_sock.cc [hunk 1]
```diff
@@ -139,12 +149,25 @@ static bool ValidateNetlink(char *data) {
         assert(0);
         return false;
     }
-#endif
+
```
Comment (mclapinski) [inline src/ksync/ksync_sock.cc]: maybe delete this line

File: src/ksync/ksync_sock.cc [hunk 2]
```diff
@@ -247,6 +270,7 @@ KSyncSock::~KSyncSock() {
     if (nl_client_->cl_buf) {
         free(nl_client_->cl_buf);
     }
+
```
Comment (mclapinski) [inline src/ksync/ksync_sock.cc]: delete this line

File: src/ksync/ksync_sock.cc [hunk 3]
```diff
@@ -622,44 +716,55 @@ bool KSyncSockNetlink::Decoder(char *data, AgentSandeshContext *context) {
...
 }
 
+#ifndef _WINDOWS
 void KSyncSockNetlink::AsyncReceive(mutable_buffers_1 buf, HandlerCb cb) {
-    //WINDOWSFIX sock_.async_receive(buf, cb);
+    sock_.async_receive(buf, cb);
 }
+#else
+void KSyncSockNetlink::AsyncReceive(mutable_buffers_1 buf, HandlerCb cb) {
+    pipe_.async_read_some(buf, cb);
```
Comment (mclapinski) [inline src/ksync/ksync_sock.cc]: maybe add comment with todo and task number here

File: src/ksync/ksync_sock.h [hunk 4]
```diff
@@ -358,7 +358,11 @@ class KSyncSock {
 //netlink socket class for interacting with kernel
 class KSyncSockNetlink : public KSyncSock {
 public:
+#ifdef _WINDOWS
```
Comment (mclapinski) [inline src/ksync/ksync_sock.h]: Maybe switch to #ifndef in headers

File: src/vnsw/agent/vrouter/ksync/ksync_init.cc [hunk 5]
```diff
@@ -355,8 +358,11 @@ void GenericNetlinkInit() {
     LOG(DEBUG, "Vrouter family is " << family);
     KSyncSock::SetNetlinkFamilyId(family);
     nl_free_client(cl);
+#else
+    // Value is unused under Windows, but method
+    // initializes internal structures
+    KSyncSock::SetNetlinkFamilyId(0);
```
Comment (mclapinski) [inline src/vnsw/agent/vrouter/ksync/ksync_init.cc:360]: If using `family` variable doesn't break anything then please stick with it.

**Review by mclapinski — CHANGES_REQUESTED** [review]

File: src/ksync/ksync_sock.cc [hunk 6]
```diff
@@ -80,25 +82,33 @@ static uint32_t IoVectorToData(char *data, KSyncBufferList *iovec) {
...
+#ifndef _WINDOWS
 static uint32_t GetNetlinkSeqno(char *data) {
-#if 0 //WINDOWSFIX
     struct nlmsghdr *nlh = (struct nlmsghdr *)data;
     return nlh->nlmsg_seq;
-#endif
-	return 0;
 }
+#else
+static uint32_t GetNetlinkSeqno(char *data) {
```
Comment (sodar) [inline src/ksync/ksync_sock.cc:91]: Redundant `GetNetlinkSeqno` definitions could be removed by moving `ifdef` inside the function. What I mean is something like this:

```c
static uint32_t GetNetlinkSeqno(char *data) {
#ifndef _WINDOWS
/* original code */
#else
/* code specific to Windows */
#endif
}
```

File: src/ksync/ksync_sock.cc [hunk 7]
```diff
@@ -80,25 +82,33 @@ static uint32_t IoVectorToData(char *data, KSyncBufferList *iovec) {
...
+#ifndef _WINDOWS
 static bool NetlinkMsgDone(char *data) {
-#if 0 //WINDOWSFIX
     struct nlmsghdr *nlh = (struct nlmsghdr *)data;
     return ((nlh->nlmsg_flags & NLM_F_MULTI) != 0);
-#endif
-	return false;
 }
+#else
+static bool NetlinkMsgDone(char *data) {
```
Comment (sodar) [inline src/ksync/ksync_sock.cc:103]: Same comment as above, regarding `GetNetlinkSeqno`.

File: src/ksync/ksync_sock.cc [hunk 8]
```diff
@@ -139,12 +149,24 @@ static bool ValidateNetlink(char *data) {
         assert(0);
         return false;
     }
-#endif
     return true;
 }
+#else
+static bool ValidateNetlink(char *data) {
```
Comment (sodar) [inline src/ksync/ksync_sock.cc:155]: Same comment as above, regarding `GetNetlinkSeqno`.

File: src/ksync/ksync_sock.cc [hunk 9]
```diff
@@ -155,34 +177,35 @@ static void GetNetlinkPayload(char *data, char **buf, uint32_t *buf_len) {
 
     *buf = data + len;
     *buf_len = nlh->nlmsg_len - len;
-#endif
 }
+#else
+static void GetNetlinkPayload(char *data, char **buf, uint32_t *buf_len) {
```
Comment (sodar) [inline src/ksync/ksync_sock.cc:182]: Same comment as above, regarding `GetNetlinkSeqno`.

File: src/ksync/ksync_sock.h [hunk 10]
```diff
@@ -358,7 +358,11 @@ class KSyncSock {
 //netlink socket class for interacting with kernel
 class KSyncSockNetlink : public KSyncSock {
 public:
+#ifndef _WINDOWS
     KSyncSockNetlink(boost::asio::io_service &ios, int protocol);
+#else
+    KSyncSockNetlink(boost::asio::io_service &ios);
```
Comment (sodar) [inline src/ksync/ksync_sock.h]: I think that we should leave `protocol` argument in constructor and other methods declarations. It won't be used, but less of the original code is changed this way. If compiler complains about unreferenced parameter, we could put `protocol` in some sort of `UNUSED` macro.

File: src/ksync/ksync_sock.h [hunk 11]
```diff
@@ -374,9 +378,18 @@ class KSyncSockNetlink : public KSyncSock {
 
     static void NetlinkDecoder(char *data, SandeshContext *ctxt);
     static void NetlinkBulkDecoder(char *data, SandeshContext *ctxt, bool more);
+#ifndef _WINDOWS
     static void Init(boost::asio::io_service &ios, int protocol);
+#else
+    static void Init(boost::asio::io_service &ios);
```
Comment (sodar) [inline src/ksync/ksync_sock.h]: Same comment as to `KsyncSockNetlink` constructor.

File: src/vnsw/agent/vrouter/ksync/ksync_init.cc [hunk 12]
```diff
@@ -125,17 +125,20 @@ void KSync::InitFlowMem() {
...
     EventManager *event_mgr;
 
     event_mgr = agent_->event_manager();
     boost::asio::io_service &io = *event_mgr->io_service();
 
-    //WINDOWFIX KSyncSockNetlink::Init(io, NETLINK_GENERIC);
+#ifdef _WINDOWS
+    KSyncSockNetlink::Init(io);
+#else
+    KSyncSockNetlink::Init(io, NETLINK_GENERIC);
```
Comment (sodar) [inline src/vnsw/agent/vrouter/ksync/ksync_init.cc]: If `KsyncSockNetlink::Init` definition changes, then conditional compilation can be removed from this part.

File: src/ksync/ksync_sock.cc [hunk 13]
```diff
@@ -554,16 +600,51 @@ KSyncSockNetlink::KSyncSockNetlink(boost::asio::io_service &ios, int protocol)
...
+KSyncSockNetlink::KSyncSockNetlink(boost::asio::io_service &ios)
+    : pipe_(ios)
+{
+    DWORD access_flags = GENERIC_READ | GENERIC_WRITE;
+    DWORD attrs = OPEN_EXISTING;
+
+    nl_client_->cl_win_pipe = CreateFile(KSYNC_PATH, access_flags, 0, NULL, attrs, 0, NULL);
+    if (nl_client_->cl_win_pipe == INVALID_HANDLE_VALUE) {
+        LOG(ERROR, "Error while opening KSync pipe: " << GetFormattedWindowsErrorMsg());
+        assert(0);
```
Comment (sodar) [inline src/ksync/ksync_sock.cc:614]: I think these `asserts` are misused here. Opening the file can fail for other reasons than a programmer error (e.g. extension was not enabled). When project is compiled using release settings, assert macros are compiled to empty instructions, thus an agent will not terminate. I think `exit` or `abort` should be used here.

File: src/ksync/ksync_sock.cc [hunk 14]
```diff
@@ -222,13 +244,13 @@ KSyncSock::KSyncSock() :
         sprintf(name, "KSync Receive Queue-%d", i);
         receive_work_queue[i]->set_name(name);
     }
+
```
Comment (mclapinski) [inline src/ksync/ksync_sock.cc:247]: pay more attention to that next time

**Review by mclapinski — APPROVED** [review]

File: src/ksync/ksync_sock.cc [hunk 15]
```diff
@@ -554,16 +600,51 @@ KSyncSockNetlink::KSyncSockNetlink(boost::asio::io_service &ios, int protocol)
...
+    if (nl_client_->cl_win_pipe == INVALID_HANDLE_VALUE) {
+        LOG(ERROR, "Error while opening KSync pipe: " << GetFormattedWindowsErrorMsg());
+        assert(0);
+    }
+
+    boost::system::error_code ec;
+    pipe_.assign(nl_client_->cl_win_pipe, ec);
+    if (ec) {
+        LOG(ERROR, "Error while assigning KSync pipe: " << ec);
+        assert(0);
```
Comment (sodar) [inline src/ksync/ksync_sock.cc:621]: Same comment as above regarding `assert`.

File: src/ksync/ksync_sock.cc [hunk 16]
```diff
@@ -554,16 +600,51 @@ KSyncSockNetlink::KSyncSockNetlink(boost::asio::io_service &ios, int protocol)
...
+}
+#endif
 
 KSyncSockNetlink::~KSyncSockNetlink() {
+#ifdef _WINDOWS
+    boost::system::error_code ec;
+    pipe_.close(ec);
+    if (ec) {
+        LOG(ERROR, "Error closing KSync pipe: " << ec);
+        assert(0);
```
Comment (sodar) [inline src/ksync/ksync_sock.cc:632]: Same comment regarding asserts.

File: src/ksync/ksync_sock.cc [hunk 17]
```diff
@@ -599,19 +682,28 @@ size_t KSyncSockNetlink::SendTo(KSyncBufferList *iovec, uint32_t seq_no) {
...
 
-    //WINDOWSFIX boost::asio::netlink::raw::endpoint ep;
-	return 0;//WINDOWSFIX sock_.send_to(*iovec, ep);
+#ifndef _WINDOWS
+    boost::asio::netlink::raw::endpoint ep;
+    return sock_.send_to(*iovec, ep);
+#else
+    boost::system::error_code ec;
+    size_t sz = boost::asio::write(pipe_, *iovec, ec);
+    if (ec) {
```
Comment (sodar) [inline src/ksync/ksync_sock.cc]: Does `send_to` throws an exception on failure? If it does, does agent code expects this exception to be thrown somewhere?

File: src/ksync/ksync_sock.cc [hunk 18]
```diff
@@ -622,44 +714,55 @@ bool KSyncSockNetlink::Decoder(char *data, AgentSandeshContext *context) {
...
     }
-#endif
 }
+#else
+void KSyncSockNetlink::Receive(mutable_buffers_1 buf) {
+    boost::system::error_code ec;
+    pipe_.read_some(buf, ec); // TODO: JW-408
+    if (ec) {
+        LOG(ERROR, "Pipe read failed: " << ec);
+        assert(0);
```
Comment (sodar) [inline src/ksync/ksync_sock.cc]: Same comment regardng use of `asserts`.

**Review by sodar — CHANGES_REQUESTED** [review]

**Comment by jablonskim** [comment]
Retest this please

**Review by sodar — APPROVED** [review]

**Pull request closed without merging**